### PR TITLE
channeldb: splite `UpdateInvoice` logic

### PR DESCRIPTION
**Pull request by positiveblue**
The `InvoiceDB` interface has one method a method that takes care of updating an invoice. An invoice can be updated in multiple ways, it is up to the caller to decide how an invoice need to be updated and persisted using one of the function arguments.

The method will load the current invoice state from the db and apply a callback provided by the user. The callback needs to return an `InvoiceUpdateDesc`. Depending on the kind of update, the `InvoiceUpdateDesc` will be populated with information that will be used to update the db entry. Because `UpdateInvoice` is used in multiple scenarios (cancel an htlc, settle an invoice, cancel an invoice etc...) the `InvoiceUpdateDesc` bundles parameters used in each type of update. No update will ever use all the fields of the struct.

Because the flow for every kind of update + invoice type (bolt11 vs AMP) is mixed it is hard to follow what constraints the `InvoiceUpdateDesc` fields have + what parts get modified in every type of call.

The objective of this PR is to split  the logic of each type of update that is allowed. This would allow keep better track of the each update flow, logically decouple the flows of each update type and possibly prepare the path to split this method in multiple ones.

Now that the InvoiceDB interface is exposed and other people can code and plug their own implementations it may make sense to have a clear API to what needs to be persisted in each method while keeping the checks of what is allowed in the `invoices` package.

*NOTE:* things to take into account
- We never add and cancel htlcs in the same call.
- We never add/cancel more than once htlc in one call. For AMP invoices though cancelling one htlc may trigger canceling all the htlcs with the same SetID.
- We never add/cancel htlcs belonging to more than one SetID. 
- Adding an HTLC may trigger settleing an invoice. State changes are part of the data bundled in `InvoiceUpdateDesc` so the callback needs to take care of setting them.
- ContractAccepted is only used in hold invoices

### CODE REVIEW

**Review by bitromortac — COMMENTED** [review]
I see the necessity to break this up and make the current invoice update handling more separated :100:. It feels quite difficult to get all things right at once here, especially with AMP in the mix. Could one extract AMP code paths perhaps in their own functions as well?

File: invoices/invoices.go [hunk 1]
```diff
@@ -656,6 +656,50 @@ type HtlcAcceptDesc struct {
 	AMP *InvoiceHtlcAMPData
 }
 
+// UpdateType is an enum that describes the type of update that was applied to
+// an invoice.
+type UpdateType uint8
```
Comment (sputn1ck) [inline invoices/invoices.go:661]: Q: Why wouldn't we want to add explicit structs for the corresponding updates? E.g. `CancelHtlcsUpdate` `SettleHodlInvoiceUpdate`

**Review by sputn1ck — COMMENTED** [review]
Easy to follow PR 🥇 Still trying to grok some of what is happening. Have a Q regarding using explicit types.

File: channeldb/invoices.go [hunk 2]
```diff
@@ -2123,91 +2216,14 @@ func (d *DB) updateInvoice(hash *lntypes.Hash, refSetID *invpkg.SetID, invoices,
...
-	}
-
-	return &invoice, nil
-}
-
-// cancelHTLCs tries to cancel the htlcs in the given InvoiceUpdateDesc.
-//
-// NOTE: cancelHTLCs updates will only use the `CancelHtlcs` field in the
-// InvoiceUpdateDesc.
-func (d *DB) cancelHTLCs(invoices kvdb.RwBucket, invoiceNum []byte,
```
[the remark this answers is not in the document]
Comment (positiveblue) [reply channeldb/invoices.go]: Jesus christ that looked horrible! I split that commit in two: one adding a dispatcher for the `addHTLC` type and one deleting the code for the `updateInvoice` function.

Now it's much easier to follow :pray:

File: channeldb/invoices.go [hunk 3]
```diff
@@ -2225,6 +2231,83 @@ func (d *DB) cancelHTLCs(invoices kvdb.RwBucket, invoiceNum []byte,
 	return invoice, nil
 }
 
+// settleHodlInvoice marks a hodl invoice as settled.
+//
+// NOTE: Currently it is not possible to have HODL AMP invoices.
+func (d *DB) settleHodlInvoice(invoices, settleIndex kvdb.RwBucket,
```
Comment (bitromortac) [inline channeldb/invoices.go:2253]: It looks like `settleHodlInvoice` and `cancelInvoice` don't have test coverage, not sure if needed?

**Review by bitromortac — COMMENTED** [review]
This looks close to me :rocket:, nice commit restructures!

File: channeldb/invoices.go [hunk 4]
```diff
@@ -2308,6 +2308,70 @@ func (d *DB) settleHodlInvoice(invoices, settleIndex kvdb.RwBucket,
...
+	for _, htlc := range invoice.Htlcs {
+		_, err := updateHtlc(
+			timestamp, htlc, invpkg.ContractCanceled, setID,
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Reserialize and update invoice.
```
[the remark this answers is not in the document]
Comment (positiveblue) [reply channeldb/invoices.go]: Added :pray:

File: channeldb/invoices.go (hunk 3 above)
Comment (positiveblue) [reply channeldb/invoices.go:2259]: They are not called directly but they are both part of multiple flows tested in [`invoices/invoiceregistry_test.go`](https://github.com/lightningnetwork/lnd/blob/master/invoices/invoiceregistry_test.go)


[`settleHodlInvoice`](https://github.com/lightningnetwork/lnd/blob/master/invoices/invoiceregistry_test.go#L340)
[`cancelInvoice`](https://github.com/lightningnetwork/lnd/blob/master/invoices/invoiceregistry_test.go#L326)

File: invoices/invoices.go [hunk 5]
```diff
@@ -656,6 +656,51 @@ type HtlcAcceptDesc struct {
...
+	// CancelHTLCsUpdate indicates that this update cancels one or more
+	// HTLCs.
+	CancelHTLCsUpdate
+
+	// AddHTLCsUpdate indicates that this update adds one or more HTLCs.
+	AddHTLCsUpdate
+
+	// SettleHodlInvoiceUpdate indicates that this update settles one or
+	// more HTLCs from a hodl invoice.
+	SettleHodlInvoiceUpdate
```
Comment (yyforyongyu) [inline invoices/invoices.go:677]: Haven't read the full PR, just one more state pops into my head.  Maybe there needs to be a `SettleHTLCsUpdate`? Since we have `CancelHTLCsUpdate` vs `CancelInvoiceUpdate`.

File: invoices/invoices.go [hunk 6]
```diff
@@ -656,6 +656,51 @@ type HtlcAcceptDesc struct {
 	AMP *InvoiceHtlcAMPData
 }
 
+// UpdateType is an enum that describes the type of update that was applied to
+// an invoice.
+type UpdateType uint8
+
+const (
```
Comment (yyforyongyu) [inline invoices/invoices.go:663]: 👍

File: channeldb/invoices.go [hunk 7]
```diff
@@ -2225,6 +2231,83 @@ func (d *DB) cancelHTLCs(invoices kvdb.RwBucket, invoiceNum []byte,
...
+	// can remove it from the API and set it here directly.
+	switch {
+	case update.State == nil:
+		fallthrough
+
+	case update.State.NewState != invpkg.ContractSettled:
+		fallthrough
+
+	case update.State.Preimage == nil:
+		return nil, fmt.Errorf("unable to settle hodl invoice: "+
```
Comment (yyforyongyu) [inline channeldb/invoices.go:2278]: Maybe we could move this error to be returned in the above `case update.State.NewState != invpkg.ContractSettled:` and instead return an error to indicate the `Preimage` is nil, just thinking of being explicit.

File: channeldb/invoices.go [hunk 8]
```diff
@@ -2231,6 +2237,255 @@ func (d *DB) cancelHTLCs(invoices kvdb.RwBucket, invoiceNum []byte,
 	return invoice, nil
 }
 
+// addHTLCs tries to add the htlcs in the given InvoiceUpdateDesc.
+func (d *DB) addHTLCs(invoices, settleIndex, setIDIndex kvdb.RwBucket, //nolint:lll,funlen
+	invoiceNum []byte, invoice *invpkg.Invoice, hash *lntypes.Hash,
+	update *invpkg.InvoiceUpdateDesc) (*invpkg.Invoice, error) {
+
+	var setID *[32]byte
+	invoiceIsAMP := invoice.IsAMP()
```
Comment (yyforyongyu) [inline channeldb/invoices.go:2018]: Just thinking out of the box...I wonder if it's easier to create two functions, one for adding htlcs for AMP `addHTLCsAMP`, and the other for MPP, since they have a different logical flow.

File: channeldb/invoices.go [hunk 9]
```diff
@@ -1913,250 +1913,11 @@ func (d *DB) updateInvoice(hash *lntypes.Hash, refSetID *invpkg.SetID, invoices,
 		return d.cancelInvoice(
 			invoices, invoiceNum, &invoice, hash, update,
 		)
-	}
```
Comment (yyforyongyu) [inline channeldb/invoices.go]: Happy deleting!!!

File: channeldb/invoices.go [hunk 10]
```diff
@@ -1913,11 +1918,6 @@ func (d *DB) updateInvoice(hash *lntypes.Hash, refSetID *invpkg.SetID, invoices,
 	// call back.
 	if update.State != nil {
 		setID = update.State.SetID
-	} else if update.SetID != nil {
-		// When we go to cancel HTLCs, there's no new state, but the
-		// set of HTLCs to be cancelled along with the setID affected
-		// will be passed in.
-		setID = (*[32]byte)(update.SetID)
```
Comment (yyforyongyu) [inline channeldb/invoices.go:1911]: Need to refresh my memory on AMP, meanwhile it seems the `update.SetID` and `update.State.SetID` are different? Guss my question is, the branch `update.State == nil && update.SetID != nil` doesn't seem to be reflected in `cancelInvoice`.

**Review by yyforyongyu — CHANGES_REQUESTED** [review]
First round looks good. Very nice commit structure, easy to follow👍 So happy we are refactoring this part of the code, which makes the following fix in #7463 easier🙏
Left some comments and questions. Will read more about AMP and then come for a second round!

File: channeldb/invoices.go [hunk 11]
```diff
@@ -2186,6 +2147,84 @@ func (d *DB) updateInvoice(hash *lntypes.Hash, refSetID *invpkg.SetID, invoices,
 	return &invoice, nil
 }
 
+// cancelHTLCs tries to cancel the htlcs in the given InvoiceUpdateDesc.
+//
+// NOTE: cancelHTLCs updates will only use the `CancelHtlcs` field in the
+// InvoiceUpdateDesc.
```
[the remark this answers is not in the document]
Comment (positiveblue) [reply channeldb/invoices.go]: Good catch! It is true that before the code had support to add and cancel htlcs in the same update operation (even when it was never  used). I deleted the use of `AddHtlcs` in this function :+1:

File: channeldb/invoices.go (hunk 10 above)
Comment (positiveblue) [reply channeldb/invoices.go:1911]: Yes, basically it is how it is set by the callers (the API is not even validated). So when we cancel an htlc for an amp invoice we set it in the top level, while for the rest of operations we set it inside the `State`.

File: channeldb/invoices.go (hunk 8 above)
Comment (positiveblue) [reply channeldb/invoices.go:2018]: Incoming in the next PR, I did it this way so this PR did not need to include changes in the tests (easier to review + build knowledge about the related code)

File: channeldb/invoices.go [hunk 12]
```diff
@@ -2225,6 +2231,83 @@ func (d *DB) cancelHTLCs(invoices kvdb.RwBucket, invoiceNum []byte,
...
 }
 
+// settleHodlInvoice marks a hodl invoice as settled.
+//
+// NOTE: Currently it is not possible to have HODL AMP invoices.
+func (d *DB) settleHodlInvoice(invoices, settleIndex kvdb.RwBucket,
+	invoiceNum []byte, invoice *invpkg.Invoice, hash *lntypes.Hash,
+	update *invpkg.InvoiceUpdateDesc) (*invpkg.Invoice, error) {
+
+	if invoice.IsAMP() {
```
[the remark this answers is not in the document]
Comment (positiveblue) [reply channeldb/invoices.go]: Added!

File: channeldb/invoices.go [hunk 13]
```diff
@@ -2225,6 +2231,83 @@ func (d *DB) cancelHTLCs(invoices kvdb.RwBucket, invoiceNum []byte,
...
+	var amtPaid lnwire.MilliSatoshi
+	for _, htlc := range invoice.Htlcs {
+		_, err := updateHtlc(
+			timestamp, htlc, invpkg.ContractSettled, nil,
+		)
+		if err != nil {
+			return nil, err
+		}
+
+		if htlc.State == invpkg.HtlcStateSettled {
```
[the remark this answers is not in the document]
Comment (positiveblue) [reply channeldb/invoices.go:2314]: Now I understand the `Maybe there needs to be a SettleHTLCsUpdate?` We can add it in the next PR

File: channeldb/invoices.go (hunk 3 above)
Comment (bitromortac) [reply channeldb/invoices.go:2254]: Ah, thank you, I think I confused something. A working command to see the coverage is `go test -coverprofile=coverage.out -cover -coverpkg=./... ./invoices`. In this case one needs to know in which package it's tested beforehand, so it's better to work with `make unit-cover` for these checks.

**Review by bitromortac — APPROVED** [review]
LGTM :tada:

File: channeldb/invoices.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply channeldb/invoices.go]: Since it's a behavior change rather than pure refactoring, could you add some commit messages to explain why it's ok to remove this check?
```go
// Consistency check to verify that there is no overlap between
// the add and cancel sets.
if _, added := update.AddHtlcs[key]; added {
	return nil, fmt.Errorf("added htlc %v canceled", key)
}
```

File: channeldb/invoices.go [hunk 14]
```diff
@@ -1969,13 +1969,8 @@ func (d *DB) cancelHTLCs(invoices kvdb.RwBucket, invoiceNum []byte,
...
 	}
 
-	// Reserialize and update invoice.
-	var buf bytes.Buffer
-	if err := serializeInvoice(&buf, invoice); err != nil {
-		return nil, err
-	}
-
-	if err := invoices.Put(invoiceNum, buf.Bytes()); err != nil {
+	err := d.serializeAndStoreInvoice(invoices, invoiceNum, invoice)
```
Comment (yyforyongyu) [inline channeldb/invoices.go:1972]: non-blocking: are there any reasons that we call this `serializeAndStoreInvoice` at multiple places? Instead of using it in `cancelHTLCs`, `addHTLCs` and `settleHodlInvoice` separately, I think we could just call it outside the `switch update.UpdateType {`, and return `d.serializeAndStoreInvoice`, seems like a few more lines were saved and fewer params used.

Or, do you plan to further make these actions into independent db tx(which is nice IMO)?

File: channeldb/invoices.go [hunk 15]
```diff
@@ -1893,17 +1893,20 @@ func (d *DB) updateInvoice(hash *lntypes.Hash, refSetID *invpkg.SetID, invoices,
...
-		setID    *[32]byte
+		setID *[32]byte
 	)
 
 	// We can either get the set ID from the main state update (if the
 	// state is changing), or via the hint passed in returned by the update
 	// call back.
 	if update.State != nil {
 		setID = update.State.SetID
-		newState = update.State.NewState
```
Comment (yyforyongyu) [inline channeldb/invoices.go:1906]: nit: this seems to be the wrong commit to remove it, was it because it was previously used in `cancelSingleHtlc` and now the compiler is complaining if it's not removed?

File: invoices/invoiceregistry.go [hunk 16]
```diff
@@ -664,21 +664,17 @@ func (i *InvoiceRegistry) cancelSingleHtlc(invoiceRef InvoiceRef,
 		// Lookup the current status of the htlc in the database.
 		var (
 			htlcState HtlcState
-			setID     *SetID
```
Comment (yyforyongyu) [inline invoices/invoiceregistry.go:667]: Looking more closely, I think this deletion belongs to the `channeldb: split cancelHTLCs logic in the UpdateInvoice method` commit, or an independent commit.

File: channeldb/invoices.go (hunk 10 above)
Comment (yyforyongyu) [reply channeldb/invoices.go:1911]: actually what is the "top level"? `update.State` or `update.SetID`?

I think with the deletion made in `cancelSingleHtlc`, both the  `update.State` and`update.SetID` will always be nil.

File: docs/release-notes/release-notes-0.16.0.md [hunk 17]
```diff
@@ -450,6 +450,10 @@ in the lnwire package](https://github.com/lightningnetwork/lnd/pull/7303)
   watchtower client (`--wtclient.active`) in the future. The new flag can not
   be set at all if the `--wtclient.active` flag has been set.
 
+* Split channeldb [`UpdateInvoice`
```
Comment (yyforyongyu) [inline docs/release-notes/release-notes-0.16.0.md:453]: this belongs to `0.17.0` now?

File: channeldb/invoice_test.go [hunk 18]
```diff
@@ -2156,6 +2165,10 @@ func getUpdateInvoiceAMPSettle(setID *[32]byte, preimage [32]byte,
 		}
 
 		update := &invpkg.InvoiceUpdateDesc{
+			// TODO(positiveblue): this would be an invalid update
+			// because tires to settle an AMP invoice without addig
```
Comment (yyforyongyu) [inline channeldb/invoice_test.go:2169]:
```suggestion
			// because tires to settle an AMP invoice without adding
```

File: channeldb/invoice_test.go [hunk 19]
```diff
@@ -2276,12 +2289,16 @@ func testUpdateHTLCPreimages(t *testing.T, test updateHTLCPreimageTestCase) {
 		invoice *invpkg.Invoice) (*invpkg.InvoiceUpdateDesc, error) {
 
 		update := &invpkg.InvoiceUpdateDesc{
+			// TODO(positiveblue): this would be an invalid update
+			// because tires to settle an AMP invoice without addig
```
Comment (yyforyongyu) [inline channeldb/invoice_test.go:2293]:
```suggestion
			// because tires to settle an AMP invoice without adding
```

File: channeldb/invoices.go (hunk 8 above)
Comment (yyforyongyu) [reply channeldb/invoices.go:2013]: > I did it this way so this PR did not need to include changes in the tests

what did you mean by `this way`? probably should review the next PR now...😂

File: channeldb/invoices.go (hunk 10 above)
Comment (yyforyongyu) [reply channeldb/invoices.go:1911]: oh nvm i saw it's all removed

**Review by yyforyongyu — CHANGES_REQUESTED** [review]
Left some questions/comments, once addressed this is good to go⛵️

File: channeldb/invoices.go (hunk 8 above)
Comment (positiveblue) [reply channeldb/invoices.go:2013]: Not changing the interface

File: channeldb/invoices.go [hunk 20]
```diff
@@ -2219,6 +2225,89 @@ func (d *DB) cancelHTLCs(invoices kvdb.RwBucket, invoiceNum []byte,
...
+func (d *DB) settleHodlInvoice(invoices, settleIndex kvdb.RwBucket,
+	invoiceNum []byte, invoice *invpkg.Invoice, hash *lntypes.Hash,
+	update *invpkg.InvoiceStateUpdateDesc) (*invpkg.Invoice, error) {
+
+	switch {
+	case !invoice.HodlInvoice:
+		return nil, fmt.Errorf("unable to settle hodl invoice: %v is "+
+			"not a hodl invoice", invoice.AddIndex)
+
+	case invoice.IsAMP():
```
[the remark this answers is not in the document]
Comment (positiveblue) [reply channeldb/invoices.go:2263]: right, we can delete the `IsAMP` case

File: channeldb/invoices.go (hunk 14 above)
Comment (positiveblue) [reply channeldb/invoices.go:1972]: Each one will be its own db interface method

**Review by yyforyongyu — APPROVED** [review]
LGTM🎉 Confirmed the itest failures are flake.

**Comment by yyforyongyu** [comment]
One more question: should this be based on `0-17-0-staging` branch instead of `master`?
